File: models/Comment.js
const mongoose = require('mongoose');

const commentSchema = new mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    campaign: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Campaign',
        required: true
    },
    content: {
        type: String,
        required: [true, 'Comment content is required'],
        trim: true
    },
    likes: [{
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User'
    }],
    replies: [{
        user: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User'
        },
        content: String,
        created_at: {
            type: Date,
            default: Date.now
        }
    }]
}, {
    timestamps: true
});

// After comment is saved, create activity and notification
commentSchema.post('save', async function(doc) {
    try {
        // Get campaign creator
        const campaign = await mongoose.model('Campaign').findById(doc.campaign);
        
        // Create activity
        await mongoose.model('Activity').create({
            user: doc.user,
            activity_type: 'comment',
            campaign: doc.campaign,
            details: {
                content: doc.content.substring(0, 100) // First 100 characters
            }
        });

        // Create notification for campaign creator
        if (campaign.creator.toString() !== doc.user.toString()) {
            await mongoose.model('Notification').create({
                user: campaign.creator,
                type: 'comment',
                message: 'Someone commented on your campaign',
                campaign: doc.campaign
            });
        }
    } catch (error) {
        console.error('Error in comment post-save hook:', error);
    }
});

module.exports = mongoose.model('Comment', commentSchema);

File: models/comment.py
from datetime import datetime
from bson import ObjectId

class Comment:
    def __init__(self, user_id, campaign_id, content):
        self.user_id = user_id
        self.campaign_id = campaign_id
        self.content = content
        self.created_at = datetime.utcnow()
        self.updated_at = datetime.utcnow()
        self.likes = []
        self.replies = []
        self.reported = False
        
    def to_dict(self):
        return {
            'user_id': self.user_id,
            'campaign_id': self.campaign_id,
            'content': self.content,
            'created_at': self.created_at,
            'updated_at': self.updated_at,
            'likes': self.likes,
            'replies': self.replies,
            'reported': self.reported
        }
    
    @staticmethod
    def from_dict(data):
        comment = Comment(
            user_id=data['user_id'],
            campaign_id=data['campaign_id'],
            content=data['content']
        )
        comment.created_at = data.get('created_at', datetime.utcnow())
        comment.updated_at = data.get('updated_at', datetime.utcnow())
        comment.likes = data.get('likes', [])
        comment.replies = data.get('replies', [])
        comment.reported = data.get('reported', False)
        return comment
